lua/entities/gmod_wire_expression2/base/optimizer.lua

```diff
@@ -0,0 +1,184 @@
+        return {"neg", instruction[2], {"add", instruction[2], instruction[3][3], instruction[4][3],
+                __instruction = true}}
+    end
+    -- (add x (neg y)) → (sub x y)
+    if instruction[4][1] == "neg" then
+        return {"sub", instruction[2], instruction[3], instruction[4][3]}
+    end
+    -- (add (neg x) y) → (sub y x)
+    if instruction[3][1] == "neg" then
+        return {"sub", instruction[2], instruction[4], instruction[3][3]}
+    end
+    return instruction
+end
+
+function peephole.sub(instruction)
+    -- (sub 0 x) → (neg x)
+    if instruction[3][1] == "literal" and instruction[3][3] == 0 then
+        return {"neg", instruction[2], instruction[4]}
+    end
+    -- (sub x 0) → x
+    if instruction[4][1] == "literal" and instruction[4][3] == 0 then return instruction[3] end
+    -- (sub (neg x) (neg y)) → (sub y x)
+    if instruction[3][1] == "neg" and instruction[4][1] == "neg" then
+        return {"sub", instruction[2], instruction[4][3], instruction[3][3]}
+    end
+    -- (sub x (neg y) → (add x y))
+    if instruction[4][1] == "neg" then
+        return {"add", instruction[2], instruction[3], instruction[4][3]}
+    end
+    -- (sub (neg x) y) → (neg (add x y))
+    if instruction[3][1] == "neg" then
+        return {"neg", instruction[2], {"add", instruction[2], instruction[3][3], instruction[4],
+                __instruction = true }}
+    end
+    return instruction
+end
+
+function peephole.mul(instruction)
+    if instruction[4][1] == "literal" and instruction[3][1] ~= "literal" then
+        instruction[3], instruction[4] = instruction[4], instruction[3]
+    end
+    -- (mul 1 x) → x
+    if instruction[3][1] == "literal" and instruction[3][3] == 1 then return instruction[4] end
+    -- (mul 0 x) → 0
+    if instruction[3][1] == "literal" and instruction[3][3] == 0 then return instruction[3] end
+    -- (mul -1 x) → (neg x)
+    if instruction[3][1] == "literal" and instruction[3][3] == -1 then
+        return {"neg", instruction[2], instruction[4]}
+    end
+    return instruction
+end
+
+function peephole.neg(instruction)
+    -- (neg (neg x)) → x
+    if instruction[3][1] == "neg" then return instruction[3][3] end
+    return instruction
+end
+
+peephole["if"] = function(instruction)
+    -- (if 1 x y) → x
+    -- (if 0 x y) → y
+    if instruction[3][1] == "literal" then
+        instruction[3] = evaluateUnary({"is", instruction[2], instruction[3]})
+        if instruction[3][3] == 1 then return instruction[4] end
+        if instruction[3][3] == 0 then return instruction[5] end
+        assert(false, "unreachable: `is` evaluation didn't return a boolean")
+    end
+    return instruction
+end
+
+peephole["whl"] = function(instruction)
+    -- (while 0 x) → (seq)
+    if instruction[3][1] == "literal" then
+        instruction[3] = evaluateUnary({"is", instruction[2], instruction[3]})
+        if instruction[3][3] == 0 then return {"seq", instruction[2]} end
+    end
```

Comment:
You missed a `return instruction` at the end of this function here.

Reply:
Wow, I miss statically-typed languages. And unit tests.

Before this gets merged I think it's worth creating an ad-hoc unit test system, maybe automatically verifying a large number of E2 scripts in a directory.